transport-classes-io_uring/src/main/java/io/netty5/incubator/channel/uring/AbstractIOUringChannel.java

```diff
@@ -0,0 +1,802 @@
+     *
+     * @return {@code true} if the channel should be closed, e.g. if a zero-readable buffer means EOF.
+     */
+    protected abstract boolean processRead(ReadSink readSink, Object read);
+
+    @Override
+    protected void readLoopComplete() {
+        super.readLoopComplete();
+        // If there are pending reads, or we received RDHUB (such that we want to drain inbound buffer),
+        // then schedule a read to run later, after other tasks.
+        // Those other tasks might issue their own reads, which we should not interferre with.
+        // Another reason we need to schedule these to run later, is that the read-loop will cancel any unprocessed
+        // reads after this method call.
+        if (isReadPending()) {
+            executor().execute(pendingRead);
+        } else if (receivedRdHub) {
+            executor().execute(rdHubRead);
+        }
+    }
+
+    @NotNull
+    protected Buffer intoDirectBuffer(Buffer buf, boolean dispose) {
+        BufferAllocator allocator = bufferAllocator();
+        assert allocator.getAllocationType() == StandardAllocationTypes.OFF_HEAP;
+        Buffer copy = allocator.allocate(buf.readableBytes());
+        copy.writeBytes(buf);
+        if (dispose) {
+            buf.close();
+        }
+        return copy;
+    }
+
+    @Override
+    protected void doWriteNow(WriteSink writeSink) throws Exception {
+        submitAllWriteMessages(writeSink);
+        // We *MUST* submit all our messages, since we'll be releasing the outbound buffers after the doWriteNow call.
+        submissionQueue.submit();
+        // Tell the write-loop to stop, but also that nothing has been written yet.
+        writeSink.complete(0, 0, 0, false);
+    }
+
+    protected abstract void submitAllWriteMessages(WriteSink writeSink);
+
+    abstract void writeComplete(int result, long udata);
+
+    /**
+     * Connect to the remote peer
+     */
+    @Override
+    protected boolean doConnect(SocketAddress remoteAddress, SocketAddress localAddress, Buffer initialData)
+            throws Exception {
+        logger.debug("doConnect: {}, remote={}, local={}, init data={}",
+                this, remoteAddress, localAddress, initialData);
+        if (localAddress instanceof InetSocketAddress) {
+            checkResolvable((InetSocketAddress) localAddress);
+        }
+
+        InetSocketAddress remoteSocketAddr = remoteAddress instanceof InetSocketAddress
+                ? (InetSocketAddress) remoteAddress : null;
+        if (remoteSocketAddr != null) {
+            checkResolvable(remoteSocketAddr);
+        }
+
+        if (localAddress != null) {
+            socket.bind(localAddress);
+        }
+
+        submitConnect(remoteSocketAddr, initialData);
+        return false;
+    }
+
+    protected void submitConnect(InetSocketAddress remoteSocketAddr, Buffer initialData) throws IOException {
+        Buffer addrBuf = bufferAllocator().allocate(Native.SIZEOF_SOCKADDR_STORAGE);
```

Comment:
What does guard us against the user to set the allocator to heap later on ?

Reply:
Good point. I'll add an accessor that always checks.